### PR TITLE
Upgrade Ensure CLI to Go 1.21+ (and support Go 1.23+)

### PR DESCRIPTION
**Pull request by JosiahWitt**
This PR adds support for Go 1.23 and, for simplicity, bumps the minimum CLI version to Go 1.21 (which is where [Go Toolchain](https://go.dev/doc/toolchain) support was added). (Technically only Go 1.22+ is supported, but Go 1.21 knows how to automatically download Go 1.22 for compilation, so it's supported.) This is in line with only the last 2 minor Go versions being officially supported, as stated in the [README](https://github.com/JosiahWitt/ensure?tab=readme-ov-file#go-version-support).

### CODE REVIEW

File: cmd/ensure/internal/ifacereader/ifacereader_test.go [hunk 1]
```diff
@@ -510,6 +510,196 @@ func buildTypeTests() []entry {
 	return entries
 }
 
+func buildGenericTests() []entry {
```
Comment (JosiahWitt) [inline cmd/ensure/internal/ifacereader/ifacereader_test.go:513]: These were just inlined from `ifacereader_go118_test.go`

File: cmd/ensure/internal/ifacereader/ifacereader.go [hunk 2]
```diff
@@ -258,6 +258,29 @@ func (r *internalPackageReader) buildTuple(variableName string, rawType types.Ty
 	}
 }
 
+func (r *internalPackageReader) parseTypeParams(namedType *types.Named) []*TypeParam {
```
Comment (JosiahWitt) [inline cmd/ensure/internal/ifacereader/ifacereader.go:261]: This was just inlined from `ifacereader_go118.go`

File: cmd/ensure/internal/ifacereader/ifacereader.go [hunk 3]
```diff
@@ -258,6 +258,29 @@ func (r *internalPackageReader) buildTuple(variableName string, rawType types.Ty
...
+	typeParamList := namedType.TypeParams()
+	typeParamCount := typeParamList.Len()
+
+	if typeParamCount == 0 {
+		return nil
+	}
+
+	typeParams := make([]*TypeParam, 0, typeParamCount)
+	for i := range typeParamCount {
```
Comment (JosiahWitt) [inline cmd/ensure/internal/ifacereader/ifacereader.go:270]: I believe the only change is the new range integer syntax, which was added in Go 1.22.

**Review by nc-treadwellj — APPROVED** [review]
I don't see much to complain about.

**Review by nc-wittj — APPROVED** [review]

**Pull request merged**